kernel/src/grant.rs

```diff
@@ -72,190 +89,384 @@ impl<'a, T: Default> AppliedGrant<'a, T> {
+                // `.enter()` closure. That is, you need to close the grant
+                // region you are currently in before trying to iterate over all
+                // grant regions.
+                panic!("Attempted to re-enter a grant region.");
+            })
+            .ok();
+
+        // # `unwrap()` Safety
+        //
+        // Only `unwrap()` if some, otherwise return early.
+        let grant_ptr = if optional_grant_ptr.is_some() {
+            optional_grant_ptr.unwrap()
+        } else {
+            return None;
+        };
+
+        // Process only holds the grant's memory, but does not know the actual type
+        // of the grant. We case the type here so that the user of the grant is restricted
+        // by the type system to access this memory safely.
+        //
+        // # Safety
+        //
+        // This is safe as long as the memory at grant_ptr is correctly aligned,
+        // the correct size for type `T`, is only every cast as a `T`, and only
+        // one reference to the object exists. We guarantee this because type
+        // `T` cannot change, and we ensure the size and alignment are correct
+        // when the grant is allocated. We ensure that only one reference can
+        // ever exist by marking the grant entered in `enter_grant()`, and
+        // subsequent calls to `enter_grant()` will fail.
+        let grant = unsafe { &mut *(grant_ptr as *mut T) };
+
+        // Setup an allocator in case the capsule needs additional memory in the
+        // grant space.
         let mut allocator = Allocator {
             appid: self.process.appid(),
         };
-        let mut root = Borrowed::new(self.grant, self.process.appid());
-        // Mark the grant region as entered by replacing its grant pointer with
-        // all 1s. This allows us to check whether the grant region is already entered.
-        unsafe {
-            self.process.set_grant_ptr(self.grant_num, !0 as *mut u8);
-        }
+
+        // Create a wrapped object that is passed to the capsule.
+        let mut root = Borrowed::new(grant);
+
+        // Allow the capsule to access the grant.
         let res = fun(&mut root, &mut allocator);
-        // Restore the grant pointer indicating that we are no longer accessing
-        // the grant.
-        unsafe {
-            self.process
-                .set_grant_ptr(self.grant_num, root.data as *mut _ as *mut u8);
-        }
-        res
+
+        // Now that the capsule has finished we need to "release" the grant.
+        // This will mark it as no longer entered and allow the grant to be used
+        // in the future.
+        self.process.leave_grant(self.grant_num);
+
+        Some(res)
     }
 }
 
-/// Grant which was dynamically allocated in a particular app's memory.
-pub struct DynamicGrant<T: ?Sized> {
-    data: NonNull<T>,
+/// Grant which was dynamically allocated from the kernel-owned grant region in
```

Comment:
It might be useful to specify that regular grants are static, or call them StaticGrant?

Reply:
I avoided that because _all_ grants are dynamically allocated, and any grant allocation (AppliedGrant or DynamicGrant) can fail due to lack of memory.

Reply:
The comment says "grant which was dynamically allocated", which implies there are grants which are **not** dynamically allocated? It seems like we need to figure out the terminology here.